pkg/util/kubernetes/autoscalers/processor_test.go

```diff
@@ -474,3 +494,97 @@ func TestInvalidate(t *testing.T) {
 		require.WithinDuration(t, time.Now(), time.Unix(e.Timestamp, 0), 5*time.Second)
 	}
 }
+
+func TestUpdateRateLimiting(t *testing.T) {
+	type Results struct {
+		Limit     float64
+		Period    float64
+		Remaining float64
+		Reset     float64
+	}
+
+	tests := []struct {
+		desc       string
+		rateLimits map[string]datadog.RateLimit
+		results    Results
+		error      error
+	}{
+		{
+			desc: "Nominal case",
+			rateLimits: map[string]datadog.RateLimit{
+				"/v1/query": {
+					Limit:     "12",
+					Period:    "3600",
+					Reset:     "11",
+					Remaining: "120",
+				},
+			},
+			results: Results{
+				Limit:     12,
+				Period:    3600,
+				Reset:     11,
+				Remaining: 120,
+			},
+			error: nil,
+		},
+		{
+			desc: "Missing header case",
+			rateLimits: map[string]datadog.RateLimit{
+				"/v1/query": {
+					Limit:  "12",
+					Period: "3600",
+					Reset:  "11",
+				},
+			},
+			results: Results{
+				Limit:  12,
+				Period: 3600,
+				Reset:  11,
+			},
+			error: fmt.Errorf("strconv.Atoi: parsing \"\": invalid syntax"),
+		},
+		{
+			desc: "Missing headers case",
+			rateLimits: map[string]datadog.RateLimit{
+				"/v1/query": {
+					Limit:  "12",
+					Period: "3600",
+				},
+			},
+			results: Results{
+				Limit:  12,
+				Period: 3600,
+			},
+			// Although several headers are missing, the Aggregate will only return 1 error as they are the same
+			error: fmt.Errorf("strconv.Atoi: parsing \"\": invalid syntax"),
+		},
+	}
+
+	for i, tt := range tests {
+		t.Run(fmt.Sprintf("#%d %s", i, tt.desc), func(t *testing.T) {
+			datadogClient := &fakeDatadogClient{
+				getRateLimitsFunc: func() map[string]datadog.RateLimit {
+					return tt.rateLimits
+				},
+			}
+			hpaCl := &Processor{datadogClient: datadogClient, externalMaxAge: maxAge}
+
+			err := hpaCl.updateRateLimiting()
+			if err != nil {
+				assert.EqualError(t, tt.error, err.Error())
+			}
+			assert.Equal(t, promutil.ToFloat64(*rateLimitsLimit), tt.results.Limit)
+			assert.Equal(t, promutil.ToFloat64(*rateLimitsReset), tt.results.Reset)
+			assert.Equal(t, promutil.ToFloat64(*rateLimitsPeriod), tt.results.Period)
+			assert.Equal(t, promutil.ToFloat64(*rateLimitsRemaining), tt.results.Remaining)
+		})
+		resetCounters(queryEndpoint)
```

Comment:
I wonder if there is a way to scope initialization to the test instead of resetting the counters after each test.

Reply:
unfortunately, the prometheus server persists the values, and they are initialised [here](https://github.com/DataDog/datadog-agent/blob/charlyf/dca-rate-limiting/pkg/util/kubernetes/autoscalers/datadogexternal.go#L72-L80). So I can't make use of the methods `Delete` or `Reset` as that would mean re-initialising the metrics.

Reply:
it think a possible solution will be to use a dedicated [prometheus.Registry](https://godoc.org/github.com/prometheus/client_golang/prometheus#NewRegistry) for each test.
but it will require some code change in the (`init()` function](https://github.com/DataDog/datadog-agent/pull/4669/files#diff-05ddad3890afefbc763f110785c0f537R72)  and in the [`NewProcessor()`](https://github.com/DataDog/datadog-agent/pull/4669/files#diff-31878728b8b72cdd470c0d6bb81f3d56R59) function

Reply:
also you should move the `rateLimitsRemaining = prometheus.NewGaugeVec` and other gauges as `Processor` members